### PR TITLE
DimValue mapping for reshape, broadcast and slice.

### PR DESCRIPTION
**Pull request by JamesTheZ**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/shape_utils.cc [hunk 1]
```diff
@@ -449,6 +491,25 @@ LogicalResult ShapeAnalysis::buildBlockDimValueMap(Block* block) {
       }
       auto& decompose = dimValueMulDecompose_[keyDimVal];
       decompose.emplace_back(std::move(decs));
+    } else if (isa<lmhlo::DynamicBroadcastInDimOp>(*op)) {
```
Comment (linearhit) [inline tao_compiler/mlir/disc/transforms/shape_utils.cc:494]: is this redundant with line 563?

File: tao_compiler/mlir/disc/transforms/shape_utils.cc [hunk 2]
```diff
@@ -497,10 +556,51 @@ LogicalResult ShapeAnalysis::buildBlockDimValueMap(Block* block) {
       DimValue result(indexCast.getResult());
 
       if (failed(mapDimValueEqual(operand, result))) {
-        return WalkResult::interrupt();
+        return WalkResult::advance();
       }
       mayCreateOrMapConstInt(operand);
       mayCreateOrMapConstInt(result);
+    } else if (isa<mhlo::DynamicReshapeOp, mhlo::DynamicBroadcastInDimOp>(op)) {
```
Comment (linearhit) [inline tao_compiler/mlir/disc/transforms/shape_utils.cc:563]: what are the rules to tell if an Op has to be put here,  for example, DynamicPad/DynamicIota? 
what will happen if they are not here?

**Review by linearhit — CHANGES_REQUESTED** [review]

File: tao_compiler/mlir/disc/transforms/shape_utils.cc (hunk 2 above)
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/shape_utils.cc:563]: An op that generates shape value or utilizes shape value should be placed here. ShapeAnalysisV2 does not use such logic any more, which we can discuss offline.

File: tao_compiler/mlir/disc/transforms/shape_utils.cc (hunk 1 above)
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/shape_utils.cc:494]: 563 is for mhlo dialect. It is lmhlo here.

File: tao_compiler/mlir/disc/transforms/shape_utils.cc (hunk 2 above)
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/shape_utils.cc:563]: If they are not here, we may miss some shape information.

**Review by linearhit — APPROVED** [review]
LGTM via offline discussion

**Pull request merged**